### PR TITLE
[ConstraintElim] Simplify `MinMaxIntrinsic`

### LINKED ISSUE #75155: Missing optimization: fold `max(b, a + 1)` to `b` when `a < b`
Alive2 proof: https://alive2.llvm.org/ce/z/V82mt-

### Description:

```c
long src(long a, long b) {
    if(a < b){
        return max(b, a + 1);
    }else{
        return 0;
    }
}
```
can be folded to:
```c
long tgt(long a, long b) {
    if(a < b){
        return b;
    }else{
        return 0;
    }
}
```
Similarly, it also applies to other cases like `a > b`, `min(b, a + 1)` and so on.

### Real-world motivation

This snippet of IR is derived from [redis/src/db.c@readGetKeys](https://github.com/redis/redis/blob/c85a9b7896b9727709b71fed414bf8a117436d71/src/db.c#L2899C14-L2899C14) (after O3 pipeline).
The example above is a reduced version. If you're interested in the original suboptimal IR and optimal IR, see also:https://godbolt.org/z/YaEYnedYW

**Let me know if you can confirm that it's an optimization opportunity, thanks.**

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch replaces min/max intrinsic with one of its operands if possible.
Alive2: https://alive2.llvm.org/ce/z/LoHfYf
Fixes #75155.

### CODE REVIEW

File: llvm/test/Transforms/ConstraintElimination/minmax.ll [hunk 1]
```diff
@@ -341,6 +341,147 @@ end:
...
+; CHECK-NEXT:    [[CMP:%.*]] = icmp slt i32 [[A]], [[B]]
+; CHECK-NEXT:    br i1 [[CMP]], label [[THEN:%.*]], label [[ELSE:%.*]]
+; CHECK:       then:
+; CHECK-NEXT:    [[ADD:%.*]] = add nsw i32 [[A]], 1
+; CHECK-NEXT:    ret i32 [[B]]
+; CHECK:       else:
+; CHECK-NEXT:    ret i32 -1
+;
+start:
+  %cmp = icmp slt i32 %a, %b
```
Comment (fhahn) [inline llvm/test/Transforms/ConstraintElimination/minmax.ll:358]: Could you also add tests with sle/ule predicates?

File: llvm/test/Transforms/ConstraintElimination/minmax.ll [hunk 2]
```diff
@@ -341,6 +341,147 @@ end:
...
+  br i1 %cmp, label %then, label %else
+then:
+  %add = add nsw i32 %a, 1
+  %max = call i32 @llvm.smax.i32(i32 %b, i32 %add)
```
Comment (fhahn) [inline llvm/test/Transforms/ConstraintElimination/minmax.ll:362]: Could you also add tests where the operands are swapped?

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 3]
```diff
@@ -1260,14 +1267,12 @@ static void generateReproducer(CmpInst *Cond, Module *M,
   assert(!verifyFunction(*F, &dbgs()));
 }
 
-static std::optional<bool> checkCondition(CmpInst *Cmp, ConstraintInfo &Info,
+static std::optional<bool> checkCondition(CmpInst::Predicate Pred, Value *A,
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1270]: Could you submit a separate PR for this NFC refactoring to reduce the diff here?

File: llvm/test/Transforms/ConstraintElimination/debug.ll [hunk 4]
```diff
@@ -11,8 +11,8 @@ define i1 @test_and_ule(i4 %x, i4 %y, i4 %z) {
 ; CHECK-NEXT: Adding 'ule %y, %z'
 ; CHECK-NEXT:  constraint: %y + -1 * %z <= 0
 
-; CHECK: Checking   %t.1 = icmp ule i4 %x, %z
-; CHECK: Condition   %t.1 = icmp ule i4 %x, %z implied by dominating constraints
+; CHECK: Checking ule i4 %x, i4 %z
```
Comment (fhahn) [inline llvm/test/Transforms/ConstraintElimination/debug.ll:14]: Could we pass in the original value if we are not optimizing a use, so we can retain the info which instruction from the input we try to simplify?

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 5]
```diff
@@ -1009,6 +1009,13 @@ void State::addInfoFor(BasicBlock &BB) {
 
     if (isa<MinMaxIntrinsic>(&I)) {
       WorkList.push_back(FactOrCheck::getInstFact(DT.getNode(&BB), &I));
+      for (Use &U : I.uses()) {
+        auto *UserI = getContextInstForUse(U);
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1013]: Is queuing the users actually needed for any of the test case? It looks like in all cases the MinMax intrinsics are in a block dominated by the relevant condition.

Solving at the uses would only be needed for cases like below, unless I am missing something? If that's the case, please add support for optimizing uses of MinMaxIntrinsics separately.

```
 define i32 @simplify_umax_val(i32 %a, i32 %b) {
 start:
   %add = add nuw i32 %a, 1
   %max = call i32 @llvm.umax.i32(i32 %b, i32 %add)
   %cmp = icmp ult i32 %a, %b
   br i1 %cmp, label %then, label %else
 then:
   ret i32 %max
 else:
   ret i32 -1
 }
```

**Review by fhahn — COMMENTED** [review]
Thanks for the patch!

It would be good to split this up into 3 parts at least

1) precommit tests
2) refactor `checkCondition` interface
3) non-strict predicate support
potentially 4) optimize MinMaxIntrinsics at uses

File: llvm/test/Transforms/ConstraintElimination/minmax.ll (hunk 1 above)
Comment (dtcxzyw) [reply llvm/test/Transforms/ConstraintElimination/minmax.ll:358]: Done.

File: llvm/test/Transforms/ConstraintElimination/minmax.ll (hunk 2 above)
Comment (dtcxzyw) [reply llvm/test/Transforms/ConstraintElimination/minmax.ll:362]: Done.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 5 above)
Comment (dtcxzyw) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1013]: Done.

**Comment by fhahn** [comment]
Thanks for the updates. It's a bit hard to see what's going on now as there are multiple commits in this PR now.

**Comment by dtcxzyw** [comment]
> Thanks for the updates. It's a bit hard to see what's going on now as there are multiple commits in this PR now.

I will combine the last two commits into one commit.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 5 above)
Comment (nikic) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1013]: I believe @fhahn's point here was that you should not add support for handling uses in this PR.

Comment (fhahn) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1013]: Yep that's what I meant, would be good to split this off to a separate PR

Comment (dtcxzyw) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp]: Done.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 6]
```diff
@@ -1349,6 +1351,47 @@ static bool checkAndReplaceCondition(
...
+static bool checkAndReplaceMinMax(MinMaxIntrinsic *MinMax, ConstraintInfo &Info,
+                                  unsigned NumIn, unsigned NumOut,
+                                  Instruction *ContextInst,
+                                  Module *ReproducerModule,
+                                  ArrayRef<ReproducerEntry> ReproducerCondStack,
+                                  DominatorTree &DT,
+                                  SmallVectorImpl<Instruction *> &ToRemove) {
+  auto ReplaceMinMaxWithOperand = [&](MinMaxIntrinsic *MinMax, bool UseLHS) {
+    // TODO: generate reproducer for min/max.
+    MinMax->replaceUsesWithIf(MinMax->getOperand(UseLHS ? 0 : 1),
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1363]: As the current version doesn't try to simplify uses, it should be possible to just use `replaceAllUsesWith` here for now?

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 7]
```diff
@@ -1349,6 +1351,47 @@ static bool checkAndReplaceCondition(
   return false;
 }
 
+static bool checkAndReplaceMinMax(MinMaxIntrinsic *MinMax, ConstraintInfo &Info,
+                                  unsigned NumIn, unsigned NumOut,
+                                  Instruction *ContextInst,
+                                  Module *ReproducerModule,
+                                  ArrayRef<ReproducerEntry> ReproducerCondStack,
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1358]: unused?

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 8]
```diff
@@ -1349,6 +1351,47 @@ static bool checkAndReplaceCondition(
   return false;
 }
 
+static bool checkAndReplaceMinMax(MinMaxIntrinsic *MinMax, ConstraintInfo &Info,
+                                  unsigned NumIn, unsigned NumOut,
+                                  Instruction *ContextInst,
+                                  Module *ReproducerModule,
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1357]: unused?

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 9]
```diff
@@ -1349,6 +1351,47 @@ static bool checkAndReplaceCondition(
...
+      ToRemove.push_back(MinMax);
+    return true;
+  };
+
+  if (auto ImpliedCondition = checkCondition(
+          MinMax->getPredicate(), MinMax->getOperand(0), MinMax->getOperand(1),
+          MinMax, Info, NumIn, NumOut, ContextInst))
+    return ReplaceMinMaxWithOperand(MinMax, *ImpliedCondition);
+  if (auto ImpliedCondition = checkCondition(
+          ICmpInst::getNonStrictPredicate(MinMax->getPredicate()),
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1388]: It's not entirely obvious why this check here is needed (or the one above isn't sufficient). 

I think it would be more explicit and probably slightly clearer if we explicitly check `(Op0 (non-strict-pred) Op1)`  and `(Op1 (non-strict-pred) Op0)` instead?

**Comment by fhahn** [comment]
Looks like this needs a rebase now

**Comment by dtcxzyw** [comment]
> Looks like this needs a rebase now

Done.

File: llvm/test/Transforms/ConstraintElimination/minmax.ll [hunk 10]
```diff
@@ -341,6 +341,271 @@ end:
   ret i32 0
 }
 
+; Test from PR75155
+define i32 @simplify_slt_smax_val(i32 %a, i32 %b) {
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/test/Transforms/ConstraintElimination/minmax.ll:345]: I am not willing to add new declarations for minmax intrinsics with i16 variants :(

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 11]
```diff
@@ -1067,6 +1067,8 @@ void State::addInfoFor(BasicBlock &BB) {
     case Intrinsic::smin:
     case Intrinsic::smax:
       WorkList.push_back(FactOrCheck::getInstFact(DT.getNode(&BB), &I));
+      WorkList.push_back(
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1070]: comment above needs updating now, as this queues for simplification now. Also, `abs` shouldn't be queued for simplification, right?

File: llvm/test/Transforms/ConstraintElimination/minmax.ll (hunk 10 above)
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/test/Transforms/ConstraintElimination/minmax.ll:345]: Fair enough, would probably be worth changing the whole file separately.

**Review by fhahn — COMMENTED** [review]
Thanks for the update!

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 11 above)
Comment (dtcxzyw) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp]: Done.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 12]
```diff
@@ -1392,6 +1396,31 @@ static bool checkAndReplaceCondition(
   return false;
 }
 
+static bool checkAndReplaceMinMax(MinMaxIntrinsic *MinMax, ConstraintInfo &Info,
+                                  unsigned NumIn, unsigned NumOut,
+                                  Instruction *ContextInst, DominatorTree &DT,
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1401]: DT unused.

Comment (nikic) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1401]: Rebase over https://github.com/llvm/llvm-project/commit/ebb853fbe5f67ab42d78de45bbcb7b32b6323fb1 please, and remove other unused params.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 13]
```diff
@@ -1392,6 +1396,31 @@ static bool checkAndReplaceCondition(
...
+                                  SmallVectorImpl<Instruction *> &ToRemove) {
+  auto ReplaceMinMaxWithOperand = [&](MinMaxIntrinsic *MinMax, bool UseLHS) {
+    // TODO: generate reproducer for min/max.
+    MinMax->replaceAllUsesWith(MinMax->getOperand(UseLHS ? 0 : 1));
+    if (MinMax->use_empty())
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp]: Done.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 12 above)
Comment (dtcxzyw) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp]: Done.

**Review by nikic — APPROVED** [review]
LGTM

**Review by fhahn — APPROVED** [review]
LGTM, thanks!
